src/bootstrap/bin/rustc.rs

```diff
@@ -115,8 +117,15 @@ fn main() {
         if let Ok(host_linker) = env::var("RUSTC_HOST_LINKER") {
             cmd.arg(format!("-Clinker={host_linker}"));
         }
-        if env::var_os("RUSTC_HOST_FUSE_LD_LLD").is_some() {
-            cmd.arg("-Clink-args=-fuse-ld=lld");
+        if let Ok(lld) = env::var("RUSTC_HOST_LLD") {
+            let lld_mode = match lld.as_str() {
+                "external" => LldMode::External,
+                "self-contained" => LldMode::SelfContained,
+                _ => panic!("Unknown host lld value {lld}"),
+            };
+            for flag in get_lld_flags(&lld_mode) {
+                cmd.arg(flag);
+            }
```

Comment:
I believe you can pass the value of `get_lld_flags(&lld_mode)` from the builder to the bin shims using environment variables so that the linking of the bootstrap lib will be avoided easily.

Reply:
Hmm, surely it can't be that easy, or why would the code otherwise even introduce `RUSTC_HOST_LLD` before? (This was pre-existing). This only has to be done when `--target` is not set (i.e. when compiling host code), and I think that cargo invokes `rustc`/`rustdoc` twice (with and without target) to distinguish this, but you can't use the environment variable only for one situation from the outside.

Or did I misunderstood and you meant something like
```rust
if let Ok(args) = env::var("ADDITIONAL_HOST_ENV") {
   cmd.env/arg(...);
}
```
? In that case, I think that we would have to come up with some env/arg encoding scheme.

Reply:
What about this:
1) We will pass host CLI args using `RUSTC_HOST_CMD_0=-Clink-args=foo` syntax.
2) We will pass host env. vars using `RUSTC_HOST_ENV_FOO=BAR` syntax.
3) The shims will scan all env vars., and if they found an env.var with this prefix, they will parse the content out and apply the given command/env. var, without knowing about any specific logic (like for LLD).

One annoying thing is that we'd need to sync the cmd counter for 1) (or use some random hash). We could pass all CLI args in a single env. var, but then we'd need to somehow encode/split them, becase the shim needs to pass them as separate `cmd.arg` calls.

Reply:
That could replace all this logic:
```rust
if let Ok(host_linker) = env::var("RUSTC_HOST_LINKER") {
            cmd.arg(format!("-Clinker={host_linker}"));
        }
        if env::var_os("RUSTC_HOST_FUSE_LD_LLD").is_some() {
            cmd.arg("-Clink-args=-fuse-ld=lld");
        }

        if let Ok(s) = env::var("RUSTC_HOST_CRT_STATIC") {
            if s == "true" {
                cmd.arg("-C").arg("target-feature=+crt-static");
            }
            if s == "false" {
                cmd.arg("-C").arg("target-feature=-crt-static");
            }
        }
```

Reply:
If there's no simple way to join-split multiple args in a single env var (is there? I don't know), then `RUSTC_HOST_CMD_0` ... `RUSTC_HOST_CMD_N` seems fine to me.
Probably better done in a separate PR though, together with the migration for all this stuff https://github.com/rust-lang/rust/pull/116278#discussion_r1346951685.

Reply:
My point is that instead of sending the lld mode to the shims and handling the logic there, we can handle the logic directly in the builder and send the flags. This way, all the shims have to do is parse and use them without depending on the bootstrap libs.

Reply:
Ok, then we have understood each other :laughing: I'll first send a PR with this generalized host flag handling to clean this up.